[BUG] Incremental builds fail: unknown "Deflate::Deflate" target

**Describe the bug**

When I have OIIO find and build missing TIFF and libdeflate libraries once, there's no problem. However, if I use that same build directory for successive builds, the build system is able to find TIFF along with the other dependencies OIIO had built previously, as expected; but the build fails very shortly after, due to the TiffTargets.cmake 


**OpenImageIO version and dependencies**
building with CMake-3.29+
```
OIIO 2.6.6.0dev | Linux/ARM
    Build compiler: gcc 10.2 | C++17/201703
    HW features enabled at build: neon
    No CUDA support (disabled / unavailable at build time)
Dependencies: BZip2 NONE, DCMTK NONE, FFmpeg NONE, fmt 10.2.1, Freetype 2.13.2, GIF NONE, Imath 3.1.10, JPEG 80, JXL NONE, libdeflate 1.20, Libheif NONE, libjpeg-turbo 3.0.3, LibRaw NONE, OpenColorIO 2.3.2, OpenCV
    NONE, OpenEXR 3.2.4, OpenGL, OpenJPEG NONE, PNG 1.6.44, Ptex NONE, Ptex NONE, pybind11 2.13.6, Python3 3.9.20, Qt5 NONE, Qt6 NONE, Robinmap 1.3.0, TBB NONE, TIFF 4.6.0, WebP NONE, yaml-cpp 0.8.0, ZLIB 1.2.7
Running on 20 cores 31.4GB 
    No compute specific hardware enabled.
```
```
OIIO 2.6.3.0dev | MacOS/ARM
    Build compiler: Apple clang 15.0 | C++17/201703
    HW features enabled at build: neon
    No CUDA support (disabled / unavailable at build time)
Dependencies: BZip2 1.0.8, DCMTK NONE, FFmpeg NONE, fmt 10.2.1, Freetype 2.13.2, GIF NONE, Imath 3.1.10, JPEG 80, JXL NONE, libdeflate 1.20, Libheif
    NONE, libjpeg-turbo 3.0.3, LibRaw NONE, OpenColorIO 2.3.2, OpenCV NONE, OpenEXR 3.2.4, OpenGL, OpenJPEG NONE, PNG 1.6.43, Ptex NONE, Ptex NONE, pybind11
    2.13.5, Python3 3.9.16, Qt5 NONE, Qt6 NONE, Robinmap 1.3.0, TBB NONE, TIFF 4.6.0, WebP NONE, ZLIB 1.2.12
```

**To Reproduce**

In an environment that does not have TIFF available: 
Build OIIO twice in a row, using the same build directory


Steps to reproduce the behavior:

1. Build OIIO using a persistent build directory, like `/tmp/build_oiio`, building all missing dependencies. The build should complete successfully, and TIFF libraries should live under `/tmp/build_oiio/deps/dist/lib`.
2. Do the exact same thing again: build OIIO using the same build directory.



**Evidence**

After the first build completes successfully, the second build finds the freshly-made TIFF and libdeflate under `<build_dir>/deps/dist/lib` and elicits the following error:

```
  CMake Error at /tmp/build_oiio/deps/dist/lib/cmake/tiff/TiffTargets.cmake:61 (set_target_properties):
    The link interface of target "TIFF::tiff" contains:

      Deflate::Deflate

    but the target was not found.  Possible reasons include:

      * There is a typo in the target name.
      * A find_package call is missing for an IMPORTED target.
      * An ALIAS target is missing.

  Call Stack (most recent call first):
    /tmp/build_oiio/deps/dist/lib/cmake/tiff/TiffConfig.cmake:24 (include)
    /opt/_internal/pipx/venvs/cmake/lib/python3.10/site-packages/cmake/data/share/cmake-3.29/Modules/FindTIFF.cmake:103 (find_package)
    src/cmake/dependency_utils.cmake:374 (find_package)
    src/cmake/externalpackages.cmake:88 (checked_find_package)
    CMakeLists.txt:178 (include)
```


# 🔍 
If I'm understanding what's happening, the Deflate::Deflate alias is only created [if OIIO is building TIFF](https://github.com/AcademySoftwareFoundation/OpenImageIO/blob/fe9f7f8451415d8daef920e76f3c16423b1efe7d/src/cmake/build_TIFF.cmake#L15-L18). This isn't a problem if the build system immediately finds and uses the TIFF libraries it just built; but I guess on subsequent builds, `checked_find_package(TIFF)` immediately finds TIFF from the first build, and never has the opportunity to invoke anything in `build_TIFF.cmake`. 

This is a deceptively tricky issue, I think. I ran into this a few weeks ago, and pretty quickly discovered it wasn't as simple as lifting the `checked_find_package(libdeflate ...)` and subsequent `alias_library_if_not_exists` command to `externalpackages.cmake`, before the check for TIFF, although I can't remember what goes wrong, specifically -- just that I knew I was pretty immediately out of my depth.

# 🤷 
I've been working around this by providing `-DOpenImageIO_BUILD_LOCAL_DEPS=TIFF` to CMake, forcing OIIO to always build TIFF. This works fine.

--- a/src/cmake/Config.cmake.in
+++ b/src/cmake/Config.cmake.in
@@ -20,6 +20,16 @@ if (NOT @BUILD_SHARED_LIBS@)
     # This is required in static library builds, as e.g. PNG::PNG appears among
     # INTERFACE_LINK_LIBRARIES. If the project does not know about PNG target, it will cause
     # configuration error about unknown targets being linked in.
+    # Static libtiff configs may reference this target without importing it.
+    # https://gitlab.com/libtiff/libtiff/-/work_items/871
+    if (NOT TARGET Deflate::Deflate)
+        find_package (libdeflate CONFIG QUIET)
+        if (TARGET libdeflate::libdeflate_static)
+            add_library (Deflate::Deflate ALIAS libdeflate::libdeflate_static)
+        elseif (TARGET libdeflate::libdeflate_shared)
+            add_library (Deflate::Deflate ALIAS libdeflate::libdeflate_shared)
+        endif ()
+    endif ()
     find_dependency(TIFF)
     find_dependency(OpenColorIO)
     if (@JPEG_FOUND@)

--- a/src/cmake/externalpackages.cmake
+++ b/src/cmake/externalpackages.cmake
@@ -91,7 +91,7 @@ checked_find_package (libuhdr
                       VERSION_MIN 1.3)
 
 # Static libtiff configs may reference this target without importing it.
-# https://github.com/AcademySoftwareFoundation/OpenImageIO/issues/4439
+# https://gitlab.com/libtiff/libtiff/-/work_items/871
 if (NOT TARGET Deflate::Deflate)
     find_package (libdeflate CONFIG QUIET)
     alias_library_if_not_exists (Deflate::Deflate libdeflate::libdeflate_static)

--- a/src/cmake/externalpackages.cmake
+++ b/src/cmake/externalpackages.cmake
@@ -90,6 +90,14 @@ endif ()
 checked_find_package (libuhdr
                       VERSION_MIN 1.3)
 
+# Static libtiff configs may reference this target without importing it.
+# https://github.com/AcademySoftwareFoundation/OpenImageIO/issues/4439
+if (NOT TARGET Deflate::Deflate)
+    find_package (libdeflate CONFIG QUIET)
+    alias_library_if_not_exists (Deflate::Deflate libdeflate::libdeflate_static)
+    alias_library_if_not_exists (Deflate::Deflate libdeflate::libdeflate_shared)
+endif ()
+
 checked_find_package (TIFF REQUIRED
                       VERSION_MIN 4.0
                       RECOMMEND_MIN 4.5

--- a/src/cmake/externalpackages.cmake
+++ b/src/cmake/externalpackages.cmake
@@ -90,10 +90,14 @@ endif ()
 checked_find_package (libuhdr
                       VERSION_MIN 1.3)
 
-# Static libtiff configs may reference this target without importing it.
-# https://gitlab.com/libtiff/libtiff/-/work_items/871
+# Static libtiff configs may reference Deflate::Deflate without importing it
+# (https://gitlab.com/libtiff/libtiff/-/work_items/871), so libdeflate must be
+# located before TIFF discovery. In particular, a previously auto-built static
+# TIFF rediscovered from the local deps cache needs this; the libdeflate found
+# during build_TIFF.cmake does not carry over to later reconfigures.
 if (NOT TARGET Deflate::Deflate)
-    find_package (libdeflate CONFIG QUIET)
+    checked_find_package (libdeflate
+                          VERSION_MIN 1.18)
     alias_library_if_not_exists (Deflate::Deflate libdeflate::libdeflate_static)
     alias_library_if_not_exists (Deflate::Deflate libdeflate::libdeflate_shared)
 endif ()

